godtools/ViewControllers/Login/LoginViewController.swift

```diff
@@ -0,0 +1,138 @@
+    
+    @IBOutlet weak var usernameTextField: UITextField!
+    @IBOutlet weak var passwordTextField: UITextField!
+    
+    @IBOutlet weak var loginButton: BlueButton!
+    @IBOutlet weak var logoutButton: BlueButton!
+    
+    private let serverURL = URL(string: "https://thekey.me/cas/")!
+    private let clientId = " "
+    
+    let loginClient = TheKeyOAuth2Client()
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        if let loginClient = loginClient, !loginClient.isConfigured() {
+            loginClient.setServerURL(serverURL, clientId: clientId)
+        }
+        
+        usernameTextField.delegate = self
+        passwordTextField.delegate = self
+        
+        loginButton.setTitle("login".localized, for: .normal)
+        logoutButton.setTitle("logout".localized, for: .normal)
+        
+        let tapRecognizer = UITapGestureRecognizer(target: self, action: #selector(viewWasTapped))
+        view.addGestureRecognizer(tapRecognizer)
+        
+        setViewState()
+    }
+    
+    @IBAction func loginButtonWasPressed(_ sender: Any) {
+        processLogin()
+    }
+    
+    @IBAction func logoutButtonWasPressed(_ sender: Any) {
+        loginClient?.logout()
+        setViewState()
+    }
+    
+    @objc func viewWasTapped() {
+        self.view.endEditing(true)
+    }
+    
+    private func setViewState() {
+        guard let loginClient = loginClient else {
+            return
+        }
+        
+        let isAuthenticated = loginClient.isAuthenticated()
+        
+        logoutButton.isHidden = !isAuthenticated
+        loginButton.isHidden = isAuthenticated
+        passwordTextField.isEnabled = !isAuthenticated
+        usernameTextField.isEnabled = !isAuthenticated
+    }
+    
+    private func validateInputs(_ username: String?, _ password: String?) -> Bool {
+        guard let username = username, username.count > 0 else {
+            showAlert(withTitle: "error".localized, message: "username_required".localized)
+            return false
+        }
+        
+        guard let password = password, password.count > 0 else {
+            showAlert(withTitle: "error".localized, message: "password_required".localized)
+            return false
+        }
+        
+        return true
+    }
+    
+    fileprivate func processLogin() {
+        let username = usernameTextField.text
+        let password = passwordTextField.text
+        
+        if !validateInputs(username, password) {
+            return
+        }
+        
+        guard let loginClient = loginClient else {
+            return
+        }
+        
+        loginClient.passwordGrantLogin(for: username!, password: password!) { [unowned self] (result, error) in
+            switch result {
+            case .success:
+                self.showAlert(withTitle: "success".localized, message: "")
+                self.setViewState()
+                debugPrint(loginClient.guid())
```

Comment:
note to self, remove these `debugPrint()` lines

Reply:
I like that your using the .count on Strings here since it's now available instead of someWord.characters.count